[Detection Engine][Rules] - Adds custom highlighted fields option

yctercero opened the pull request:
## Summary

Original [PR](https://github.com/elastic/kibana/pull/152871) and changes by @RubixSolver99 .  Thanks Carson for doing the heavy lifting here! This PR cleans it up to get over the line.

Allows a user to define which fields to highlight in areas where we currently use "highlighted fields" feature. This includes:

- Exception auto populate - when adding an exception item from an alert we auto populate values
- Alert details - the alert details flyout includes a "Highlighted fields" section. 
- New alert details - the "Highlighted fields" and "Prevalence" sections

Also ensure:

- the field is populating properly in the rule details/creation/edit pages

New rule field being added is `investigation_fields`. Not adding it to the alert schemas as we will always be querying for latest in our UX, not looking to pull the custom highlighted fields from the alert doc. This means that this field will be available in the alert doc under `kibana.alert.rule.parameters["investigation_fields"]`.

### To test

- Go to rule creation
- Under About > Advanced, find "Custom highlighted fields" option and select one or more fields
- Once alerts are generated:
  - Select to view alert details and ensure the selected custom highlighted fields are showing (if there is alert data to populate with)
  - Select to view alert details and ensure the selected custom highlighted fields are showing under the "Prevelance" tab
  - Select to create an exception from alert and ensure that custom highlighted fields are included and populated with alert data 

### Updated UI screenshots
For the below screenshots, I've created a rule with a custom highlighted field of `host.os.name`
#### Rule creation
<img width="1285" alt="Screenshot 2023-08-13 at 10 20 25 PM" src="https://github.com/elastic/kibana/assets/10927944/9fda3a03-f980-458d-964a-3206f262cfcd">
<img width="1014" alt="Screenshot 2023-08-13 at 10 29 14 PM" src="https://github.com/elastic/kibana/assets/10927944/efefb7d3-cad6-4970-8af3-bd875a3ff40c">

#### Rule details
<img width="799" alt="Screenshot 2023-08-13 at 10 24 59 PM" src="https://github.com/elastic/kibana/assets/10927944/2d174f9e-0151-4ab9-a3e0-97b0f076e3cf">

#### Alerts flyout
<img width="1306" alt="Screenshot 2023-08-13 at 10 29 57 PM" src="https://github.com/elastic/kibana/assets/10927944/78ab5b03-9310-4211-98c7-ed92baeb6951">
<img width="434" alt="Screenshot 2023-08-13 at 10 29 39 PM" src="https://github.com/elastic/kibana/assets/10927944/b5abc7b7-a2dc-4587-85d7-d139109fd48b">

#### Exceptions auto-populate
<img width="1427" alt="Screenshot 2023-08-13 at 10 30 16 PM" src="https://github.com/elastic/kibana/assets/10927944/4074308a-8cd3-476b-911c-9988938b0994">


### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

yctercero committed:
trying to recreate changes for new custom highlighted fields

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
got most of the code ported over

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
adding back in tests and cleanup

yctercero committed:
continued cleanup

yctercero committed:
cleanup and making by default additive

yctercero committed:
cleaning up types

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
updating so now everywhere pulls latest custom highlighted fields

yctercero commented on x-pack/plugins/security_solution/jest.config.dev.js:
```diff
@@ -9,9 +9,9 @@ module.exports = {
   rootDir: '../../../',
   projects: [
-    '<rootDir>/x-pack/plugins/security_solution/common/*/jest.config.js',
+    // '<rootDir>/x-pack/plugins/security_solution/common/*/jest.config.js',
```
Will undo change before merging.

yctercero committed:
dont need to map it in alert since always accessing latest

yctercero committed:
continued cleanup

yctercero committed:
continued cleanup

yctercero committed:
continued cleanup

yctercero committed:
trying to fix tests

yctercero committed:
trying to fix tests

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
trying to fix tests

yctercero committed:
trying to fix tests

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
cleanup, cleanup, everybody cleanup

vitaliidm reviewed (approved):
Detection Engine code LGTM. Few minor comments from me

I played around with the functionality, it seems OK for me.

I noticed one issue, unrelated to PR

<img width="2311" alt="Screenshot 2023-08-14 at 12 13 58" src="https://github.com/elastic/kibana/assets/92328789/d66e8146-5627-41a4-92e7-95f8120d3bcb">

Numeric values in exceptions were not populated, but exist in highlighted fields. Also these highlighted fields are default, not custom.
<img width="504" alt="Screenshot 2023-08-14 at 12 17 08" src="https://github.com/elastic/kibana/assets/92328789/a0fcf5ab-51a1-434c-8dfa-efd04d97d4b9">

Is it known issue, or can we open a ticket for it?

yctercero commented:
Thanks so much for reviewing, @vitaliidm ! Addressing your comments. After talking with product, we decided that when custom highlighted fields are specified, we want it to be additive, so we will show the selected custom highlighted fields + the default highlighted fields. We may want to iterate on this in the future, but we thought this would for now address both use cases where people like the highlighted fields and also want some custom ones added.

yctercero committed:
addressing PR feedback

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_response_schema.mock.ts:
```diff
@@ -64,6 +64,7 @@ const getResponseBaseParams = (anchorDate: string = ANCHOR_DATE): SharedResponse
   timestamp_override: undefined,
   timestamp_override_fallback_disabled: undefined,
   namespace: undefined,
+  custom_highlighted_fields: undefined,
```
I think it shouldn't be possible to set it to `undefined` in the response (`RuleResponse` and other rule type-specific types). It should always be an array (empty or non-empty), which would improve DX for API users (including us on the FE side).

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_response_schema.mock.ts:
```diff
@@ -77,6 +78,7 @@ export const getRulesSchemaMock = (anchorDate: string = ANCHOR_DATE): QueryRule
   saved_id: undefined,
   response_actions: undefined,
   alert_suppression: undefined,
+  custom_highlighted_fields: undefined,
```
Same here

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_response_schema.test.ts:
```diff
@@ -232,4 +232,65 @@ describe('Rule response schema', () => {
+      expect(message.schema).toEqual(expected);
+    });
+
+    test('it should validate undefined for "custom_highlighted_fields"', () => {
+      const payload: RuleResponse = {
+        ...getRulesSchemaMock(),
+        custom_highlighted_fields: undefined,
+      };
+
+      const decoded = RuleResponse.decode(payload);
+      const checked = exactCheck(payload, decoded);
+      const message = pipe(checked, foldLeftRight);
+      const expected = { ...getRulesSchemaMock(), custom_highlighted_fields: undefined };
+
+      expect(getPaths(left(message.errors))).toEqual([]);
+      expect(message.schema).toEqual(expected);
+    });
```
Same here, validation shouldn't pass

![](https://i.kym-cdn.com/entries/icons/original/000/002/144/You_Shall_Not_Pass!_0-1_screenshot.jpg)

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.ts:
```diff
@@ -116,6 +117,7 @@ export const baseSchema = buildRuleSchemas({
     output_index: AlertsIndex,
     namespace: AlertsIndexNamespace,
     meta: RuleMetadata,
+    custom_highlighted_fields: RuleCustomHighlightedFieldArray,
```
Posting for visibility here. As discussed over zoom with @yctercero, I think this field should be:

- Defaultable instead of optional, meaning it should always be defined in `RuleResponse` and other types of rules we return from the API (`QueryRule`, `ThresholdRule`, etc).
- Wrapped in an object to make space for other parameters we might want to add for this feature in the future. This would also make it more compatible with the `DiffableRule` interface and the rule diff/upgrade algorithm.
- Renamed to something more UI-neutral and generic, e.g. "investigation fields".

So in a rule, it would be something like `rule.investigation_fields.field_names`.

We should consider writing a migration to add a "null" `investigation_fields` object to existing rules. Pseudocode:

```ts
rule.investigation_fields = {
  field_names: [],
};
```

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/description_step/helpers.tsx:
```diff
@@ -201,6 +199,38 @@ export const buildUnorderedListArrayDescription = (
   return [];
 };
 
+export const buildHighlightedFieldsOverrideDescription = (
+  label: string,
+  values: string[]
+): ListItems[] => {
+  if (isEmpty(values)) {
+    return [];
+  }
+  const description = (
+    <EuiFlexGroup responsive={false} gutterSize="xs" wrap>
+      {values.map((val: string) =>
+        isEmpty(val) ? null : (
+          <EuiFlexItem grow={false} key={`${label}-${val}`}>
+            <EuiBadgeWrap
+              data-test-subj="customHighlightedFieldsStringArrayDescriptionBadgeItem"
+              color="hollow"
+            >
+              {val}
+            </EuiBadgeWrap>
+          </EuiFlexItem>
+        )
+      )}
+    </EuiFlexGroup>
+  );
+
+  return [
+    {
+      title: label,
+      description,
+    },
+  ];
+};
```
@nikitaindik We'll need to add support for this new rule field to the components you're working on. We can do it in a separate PR once @yctercero merges this one.

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/description_step/index.test.tsx:
```diff
@@ -262,7 +262,7 @@ describe('description_step', () => {
       );
 
-      expect(result.length).toEqual(11);
+      expect(result.length).toEqual(12);
```
Interesting assertion

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/step_about_rule/index.tsx:
```diff
@@ -237,6 +238,16 @@ const StepAboutRuleComponent: FC<StepAboutRuleProps> = ({
               }}
             />
             <EuiSpacer size="l" />
+            <UseField
+              path="customHighlightedFields"
+              component={MultiSelectFieldsAutocomplete}
+              componentProps={{
+                browserFields: indexPattern.fields,
+                isDisabled: isLoading || indexPatternLoading,
+                fullWidth: true,
+              }}
+            />
```
So we do auto-suggest values, just based on the source events mappings rather than the alerts index mappings? Maybe I missed this part over zoom. The reason I'm asking is that with this auto-suggest component, it's not possible to add a field that is not in the source mappings. Is it by design?

yctercero committed:
updated jest test and linting issue

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

banderror reviewed (approved):
@yctercero has walked me through the changes over zoom. I also went through the diff myself and checked the changes that could be relevant to the Rule Management area. I also played with the feature a bit and tested that the new field works on the Rule Creation and Details pages (but please don't rely on this, it wasn't a proper manual testing).

Leaving a few comments below. I think we should address the way how this field is defined in the rule schema (see below). But I'd be ok with merging this PR and making changes in a follow-up one.

Also, the integration and e2e test coverage looks not 100% complete, could improve this aspect too.

yctercero commented on packages/kbn-securitysolution-utils/src/transform_data_to_ndjson/index.test.ts:
```diff
@@ -20,6 +20,7 @@ const getRulesSchemaMock = (anchorDate: string = ANCHOR_DATE) => ({
   enabled: true,
   false_positives: ['false positive 1', 'false positive 2'],
   from: 'now-6m',
+  custom_highlighted_fields: ['custom.field1', 'custom.field2'],
```
Agreed. Not going to touch in this PR though.

yctercero committed:
working on cypress tests and renaming field to investigation_fields

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
Merge branch 'custom_highlighted_fields' of github.com:yctercero/kibana into custom_highlighted_fields

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

PhilippeOberti reviewed (commented):
Thanks for updating the code in the new expandable flyout @yctercero, everything looks great!
One thing though, the _old_ flyout isn't reflecting the custom highlighted fields. As you can see, this screenshot shows the new flyout Highlighted Fields section (with 2 custom highlighted fields: `host.os.platform` and `@timestamp`:
![Screenshot 2023-08-15 at 10 22 32 AM](https://github.com/elastic/kibana/assets/17276605/d4031788-9cee-43bf-8e8e-39791e15462b)
But opening the same alert with the `securitySolution:enableExpandableFlyout` advanced settings turned off to show the _old_ is only showing the predefined highlighted fields:
![Screenshot 2023-08-15 at 10 22 02 AM](https://github.com/elastic/kibana/assets/17276605/ec604c8f-5ba8-48cf-bb3c-28cb240e0e65)

I feel like while we keep the `securitySolution:enableExpandableFlyout` advanced settings so that users can go back to the _old_ flyout if they want to, we should have this feature work there as well no?

e40pud reviewed (approved):
LGTM! Left couple minor comments.

yctercero committed:
adding to older event details flyout after refactor

yctercero committed:
Merge branch 'custom_highlighted_fields' of github.com:yctercero/kibana into custom_highlighted_fields

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero commented on x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/components/add_exception_flyout/index.tsx:
```diff
@@ -346,6 +346,7 @@ export const AddExceptionFlyout = memo(function AddExceptionFlyout({
           const populatedException = getPrepopulatedRuleExceptionWithHighlightFields({
             alertData,
             exceptionItemName,
+            ruleCustomHighlightedFields: rules?.[0]?.investigation_fields ?? [],
```
Removed! Thanks.

yctercero replied:
We only pre-populate when adding from an alert, in which case there's just one rule associated. But that was a good call - I can leave a comment so that that is clear here and also something to consider if we end up changing this code in the future.

yctercero commented on x-pack/plugins/security_solution/cypress/e2e/exceptions/alerts_table_flow/rule_exceptions/closing_all_matching_alerts.cy.ts:
```diff
@@ -56,7 +55,9 @@ describe('Close matching Alerts ', () => {
 
   it('Should create a Rule exception item from alert actions overflow menu and close all matching alerts', () => {
     cy.get(LOADING_INDICATOR).should('not.exist');
+    cy.log('addExceptionFromFirstAlert');
```
Removed! Thanks

yctercero commented:
@PhilippeOberti Thanks so much for the review and catching that. I've updated to now include the highlighted fields in the older alerts details view.

PhilippeOberti reviewed (approved):
Tested and code LGTM! :)

yctercero committed:
skipping flakey tests, linked to issue as required follow up for release

yctercero committed:
some cleanup

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into custom_highlighted_fields

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into custom_highlighted_fields

yctercero committed:
trying to get ci to pass

yctercero committed:
Merge branch 'custom_highlighted_fields' of github.com:yctercero/kibana into custom_highlighted_fields

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into custom_highlighted_fields

MadameSheema committed:
Merge branch 'main' into custom_highlighted_fields

mbudge commented:
It would be good if we could add fields through the bulk edit feature.

banderror commented:
@mbudge do you mind opening a feature request for this in https://github.com/elastic/kibana/issues?

mbudge commented:
https://github.com/elastic/kibana/issues/164301

mbudge commented:
My engineer is getting arthritis setting 30 custom highlighted fields in 100's of custom rules. Would be good to do this through bulk edit.

approksiu commented:
@mbudge makes sense, and bulk edit is aligned with the roadmap.
